src/GovUk.Education.ExploreEducationStatistics.Data.Model/Migrations/Current_Routine_RebuildIndexes.sql

```diff
@@ -1,120 +1,302 @@
+                    AND idx.name = @IndexName) AS Fragmented
+            WHERE __Log_RebuildIndexesAlterIndexes.RunId = @RunId
+              AND __Log_RebuildIndexesAlterIndexes.ObjectName = @ObjectName
+              AND __Log_RebuildIndexesAlterIndexes.SchemaName = @SchemaName
+              AND __Log_RebuildIndexesAlterIndexes.IndexName = @IndexName;
+
+            IF @StopTime < GETUTCDATE()
                 BEGIN
-                    SELECT @CurrentTable = ObjectName
-                    FROM #Stats
-                    WHERE Id = @StatsRowIndex;
+                    RAISERROR ('Hit timeout while processing indexes. Skipping remaining tasks', 0, 1, NULL)
+                        WITH NOWAIT;
 
-                    -- Only update statistics if we're iterating over a new table
-                    IF (@CurrentTable != @LastTable)
-                        BEGIN
-                            -- Update all statistics on the table
-                            SET @Command = 'UPDATE STATISTICS ' + @CurrentTable;
-                            RAISERROR ('Updating statistics on table ''%s''.', 0, 1, @CurrentTable) WITH NOWAIT;
-                            EXEC (@Command);
+                    -- We set the StartTime even when ActionRequired = 'None'. This allows us to update rows that
+                    -- won't be processed if we've exceeded @StopTime
+                    UPDATE __Log_RebuildIndexesAlterIndexes
+                    SET HitTimeout = 1
+                    WHERE StartTime IS NULL;
 
-                            SET @LastTable = @CurrentTable;
-                        END
+                    SET @SkipRemainingTasks = 1;
+                END
 
-                    SET @StatsRowIndex += 1;
+            SET @StatsRowIndex += 1;
+        END
+
+    RAISERROR ('Completed optimising indexes for tables.', 0, 1, NULL) WITH NOWAIT;
+
+    /*-----------------------------------------------------------------------*/
+
+    -- Call UPDATE STATISTICS on any tables that have had their indexes rebuilt/reorganized
+    SET @TablesRowIndex = 1;
+    SELECT @TablesCount = COUNT(Id) FROM #TableList;
+    WHILE @TablesRowIndex <= @TablesCount AND @SkipRemainingTasks = 0
+        BEGIN
+            SELECT @CurrentTable = ObjectName
+            FROM #TableList
+            WHERE Id = @TablesRowIndex;
+
+            SET @Command = 'UPDATE STATISTICS ' + @CurrentTable;
+
+            -- Save start time
+            UPDATE __Log_RebuildIndexesUpdateStatistics
+            SET StartTime = GETUTCDATE()
+            WHERE RunId = @RunId
+              AND TableName = @CurrentTable;
+
+            -- Skip if no indexes related to table were rebuilt/reorganized
+            SELECT @CurrentTableIndexesProcessed = COUNT(*)
+            FROM __Log_RebuildIndexesAlterIndexes
+            WHERE RunId = @RunId
+              AND ObjectName = @CurrentTable
+              AND EndTime IS NOT NULL
+
+            IF @CurrentTableIndexesProcessed = 0
```

Comment:
This could all  be replaced with

`IF EXISTS (SELECT 1 FROM __Log_RebuildIndexesAlterIndexes WHERE ...)`

which would be more efficient than using `COUNT` since it usually ends up scanning the entire table.